scripts/go_gen_p4_const.go

```diff
@@ -0,0 +1,166 @@
+
+		builder.WriteString(CTR_VAR_PREFIX + name + "\t" + UINT32_STRING + ID + "\n")
+	}
+	// Direct Counters
+	builder.WriteString("// DirectCounters\n")
+	for _, element := range p4info.GetDirectCounters() {
+		name, ID := element.GetPreamble().GetName(), strconv.FormatUint(uint64(element.GetPreamble().GetId()), 10)
+		name = strcase.ToPascal(name)
+
+		builder.WriteString(DIRCTR_VAR_PREFIX + name + "\t" + UINT32_STRING + ID + "\n")
+	}
+	// Action Param IDs
+	builder.WriteString("// ActionParams\n")
+	for _, element := range p4info.GetActions() {
+		for _, actionParam := range element.GetParams() {
+			actionName := element.GetPreamble().GetName()
+			name, ID := actionParam.GetName(), strconv.FormatUint(uint64(actionParam.GetId()), 10)
+			name = strcase.ToPascal(name)
+
+			builder.WriteString(ACTPARAM_VAR_PREFIX + actionName + name + "\t" + UINT32_STRING + ID + "\n")
+		}
+	}
+	// Action profiles
+	builder.WriteString("// ActionProfiles\n")
+	for _, element := range p4info.GetActionProfiles() {
+		name, ID := element.GetPreamble().GetName(), strconv.FormatUint(uint64(element.GetPreamble().GetId()), 10)
+
+		builder.WriteString(ACTPROF_VAR_PREFIX + name + "\t" + UINT32_STRING + ID + "\n")
+	}
+	// Packet metadata
+	builder.WriteString("// PacketMetadata\n")
+	for _, element := range p4info.GetControllerPacketMetadata() {
+		name, ID := element.GetPreamble().GetName(), strconv.FormatUint(uint64(element.GetPreamble().GetId()), 10)
+		name = strcase.ToPascal(name)
+
+		builder.WriteString(PACKETMETA_VAR_PREFIX + name + "\t" + UINT32_STRING + ID + "\n")
+	}
+	// Meters
+	builder.WriteString("// Meters\n")
+	for _, element := range p4info.GetMeters() {
+		name, ID := element.GetPreamble().GetName(), strconv.FormatUint(uint64(element.GetPreamble().GetId()), 10)
+		name = strcase.ToPascal(name)
+
+		builder.WriteString(MTR_VAR_PREFIX + name + "\t" + UINT32_STRING + ID + "\n")
+	}
+
+	builder.WriteString(CONST_CLOSE + "\n")
+
+	return builder
+}
+
+func getP4Config(p4infopath string) *p4ConfigV1.P4Info {
+	p4infoBytes, err := ioutil.ReadFile(p4infopath)
+	if err != nil {
+		panic(fmt.Sprintf("Could not read P4Info file: %v", err))
+	}
+
+	var p4info p4ConfigV1.P4Info
+
+	err = proto.UnmarshalText(string(p4infoBytes), &p4info)
+	if err != nil {
+		panic("Could not parse P4Info file")
+	}
+
+	return &p4info
+}
+
+func main() {
+	p4infoPath := flag.String("p4info", P4INFO_PATH, "Path of the p4info file")
+	outputPath := flag.String("output", "-", "Default will print to Stdout")
+	packageName := flag.String("package", DEFAULT_PACKAGE_NAME, "Set the package name")
+
+	flag.Parse()
+
+	p4config := getP4Config(*p4infoPath)
+
+	stringBuilder := new(strings.Builder)
+
+	stringBuilder.WriteString(COPYRIGHT_HEADER)
+
+	stringBuilder.WriteString(fmt.Sprintf("package %s\n", *packageName))
+	stringBuilder.WriteString(CONST_OPEN + "\n")
+
+	result := generate(p4config, stringBuilder).String()
+	result = strings.Replace(result, ".", "_", -1)
```

Comment:
The code gen logic does not seem well contained here. The start of the code is generated in main, then the constants and closing parenthesis in `generate`, lastly some replacements here again (Why?). Ideally this is one function call from main, with possible sub-functions inside it as needed.

**Code in main cannot be covered by tetsts**